File: compare_runs.mjs
#!/usr/bin/env node

import { readFile, writeFile } from "node:fs/promises";
import { statSync } from "node:fs";
import path from "node:path";

function parseArgs(argv) {
  const args = { runs: [], out: null };
  for (let index = 0; index < argv.length; index += 1) {
    const item = argv[index];
    if (item === "--out") {
      index += 1;
      if (index >= argv.length) throw new Error("Missing value after --out");
      args.out = argv[index];
    } else if (item === "--help" || item === "-h") {
      console.log("Usage: node compare_runs.mjs <run-dir|summary.json> ... [--out vergleich.md]");
      process.exit(0);
    } else {
      args.runs.push(item);
    }
  }
  if (args.runs.length === 0) throw new Error("Provide at least one run directory or summary.json file.");
  return args;
}

async function loadSummary(inputPath) {
  let summaryPath = inputPath;
  if (statSync(inputPath).isDirectory()) summaryPath = path.join(inputPath, "summary.json");
  return JSON.parse(await readFile(summaryPath, "utf8"));
}

function percent(earned, possible) {
  if (!possible) return "n/a";
  return `${Math.round((earned / possible) * 10000) / 100}%`;
}

function buildComparison(summaries) {
  const categories = [...new Set(summaries.flatMap((summary) => Object.keys(summary.categories || {})))].sort();
  const lines = [];
  lines.push("# Eval Run Comparison");
  lines.push("");
  lines.push("## Overall");
  lines.push("");
  lines.push("| Model | Run | Cases | Balanced Score | Auto Score | Failed | Errors | Avg TTFT | Avg tok/s | Avg Time |");
  lines.push("|---|---|---:|---:|---:|---:|---:|---:|---:|---:|");
  for (const summary of summaries) {
    const perf = summary.performance || {};
    lines.push(
      `| ${summary.model} | ${summary.run_id} | ${summary.case_count} | ${summary.balanced_final_percent ?? summary.balanced_auto_percent ?? balancedFromCategories(summary)}% | ${summary.balanced_auto_percent ?? balancedFromCategories(summary)}% | ${summary.status_counts?.auto_failed ?? 0} | ${summary.status_counts?.error ?? 0} | ${seconds(perf.avg_time_to_first_token_seconds)} | ${number(perf.avg_tokens_per_second)} | ${seconds(perf.avg_total_seconds)} |`,
    );
  }
  lines.push("");
  lines.push("## Categories");
  lines.push("");
  lines.push(`| Category | ${summaries.map((summary) => summary.model).join(" | ")} |`);
  lines.push(`|---|${summaries.map(() => "---:").join("|")}|`);
  for (const category of categories) {
    const cells = summaries.map((summary) => {
      const data = summary.categories?.[category] || { earned: 0, possible: 0 };
      return percent(data.final_earned ?? data.earned, data.final_possible ?? data.possible);
    });
    lines.push(`| ${category} | ${cells.join(" | ")} |`);
  }
  return `${lines.join("\n").trimEnd()}\n`;
}

function balancedFromCategories(summary) {
  const values = Object.values(summary.categories || {})
    .map((data) => {
      const earned = data.final_earned ?? data.earned;
      const possible = data.final_possible ?? data.possible;
      return possible ? (earned / possible) * 100 : null;
    })
    .filter((value) => value !== null);
  if (!values.length) return "n/a";
  return (values.reduce((sum, value) => sum + value, 0) / values.length).toFixed(2);
}

function seconds(value) {
  if (!value) return "n/a";
  return `${Number(value).toFixed(2)}s`;
}

function number(value) {
  if (!value) return "n/a";
  return Number(value).toFixed(2);
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const summaries = [];
  for (const runPath of args.runs) summaries.push(await loadSummary(runPath));
  const output = buildComparison(summaries);
  if (args.out) {
    await writeFile(args.out, output, "utf8");
    console.log(`Wrote ${args.out}`);
  } else {
    console.log(output);
  }
}

main().catch((error) => {
  console.error(`ERROR: ${error.message}`);
  process.exitCode = 1;
});

File: compare_runs.py
#!/usr/bin/env python3
"""Compare multiple run summaries produced by run_eval.py."""

from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path
from typing import Any


def load_summary(path: Path) -> dict[str, Any]:
    if path.is_dir():
        path = path / "summary.json"
    if not path.exists():
        raise FileNotFoundError(f"Missing summary: {path}")
    return json.loads(path.read_text(encoding="utf-8"))


def percent(earned: float, possible: float) -> str:
    if possible == 0:
        return "n/a"
    return f"{round(earned / possible * 100, 2)}%"


def build_comparison(summaries: list[dict[str, Any]]) -> str:
    categories = sorted({category for summary in summaries for category in summary.get("categories", {})})
    lines: list[str] = []
    lines.append("# Eval Run Comparison")
    lines.append("")
    lines.append("## Overall")
    lines.append("")
    lines.append("| Model | Run | Cases | Auto Score |")
    lines.append("|---|---|---:|---:|")
    for summary in summaries:
        lines.append(
            "| {model} | {run} | {cases} | {score} |".format(
                model=summary["model"],
                run=summary["run_id"],
                cases=summary["case_count"],
                score=percent(summary["auto_earned"], summary["auto_possible"]),
            )
        )
    lines.append("")
    lines.append("## Categories")
    lines.append("")
    header = "| Category | " + " | ".join(summary["model"] for summary in summaries) + " |"
    lines.append(header)
    lines.append("|---|" + "|".join("---:" for _ in summaries) + "|")
    for category in categories:
        cells = [category]
        for summary in summaries:
            data = summary.get("categories", {}).get(category, {"earned": 0.0, "possible": 0.0})
            cells.append(percent(data["earned"], data["possible"]))
        lines.append("| " + " | ".join(cells) + " |")
    return "\n".join(lines).rstrip() + "\n"


def parse_args(argv: list[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Compare eval run summaries.")
    parser.add_argument("runs", nargs="+", help="Run directories or summary.json files.")
    parser.add_argument("--out", help="Optional Markdown output path.")
    return parser.parse_args(argv)


def main(argv: list[str]) -> int:
    args = parse_args(argv)
    summaries = [load_summary(Path(item)) for item in args.runs]
    comparison = build_comparison(summaries)
    if args.out:
        Path(args.out).write_text(comparison, encoding="utf-8")
        print(f"Wrote {args.out}")
    else:
        print(comparison)
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
